llama-index-core/llama_index/core/indices/omni_modal/retriever.py

```diff
@@ -0,0 +1,696 @@
+                raise ValueError(
+                    "Vector store query result should return at "
+                    "least one of nodes or ids."
+                )
+            assert isinstance(self._index.index_struct, IndexDict)
+            node_ids = [
+                self._index.index_struct.nodes_dict[idx] for idx in query_result.ids
+            ]
+            nodes = self._docstore.get_nodes(node_ids)
+            query_result.nodes = nodes
+        else:
+            # NOTE: vector store keeps text, returns nodes.
+            # Only need to recover image or index nodes from docstore
+            for i in range(len(query_result.nodes)):
+                source_node = query_result.nodes[i].source_node
+                if (not vector_store.stores_text) or (
+                    source_node is not None and source_node.node_type != ObjectType.TEXT
+                ):
+                    node_id = query_result.nodes[i].node_id
+                    if self._docstore.document_exists(node_id):
+                        query_result.nodes[i] = self._docstore.get_node(node_id)  # type: ignore[index]
+
+        log_vector_store_query_result(query_result)
+
+        node_with_scores: List[NodeWithScore] = []
+        for ind, node in enumerate(query_result.nodes):
+            score: Optional[float] = None
+            if query_result.similarities is not None:
+                score = query_result.similarities[ind]
+            node_with_scores.append(NodeWithScore(node=node, score=score))
+
+        return node_with_scores
+
+    def _retrieve_from_object(
+        self,
+        obj: Any,
+        query_bundle: QueryBundle,
+        score: float,
+        *,
+        query_type: KQ,
+    ) -> List[NodeWithScore]:
+        """Retrieve nodes from object."""
+        if isinstance(obj, type(self)):
+            if self._verbose:
+                print_text(
+                    f"Retrieving from object {obj.__class__.__name__} with query {query_bundle.query_str}\n",
+                    color="llama_pink",
+                )
+
+                return obj.retrieve_multi_modal(query_bundle, query_type=query_type)
+
+        return super()._retrieve_from_object(obj, query_bundle, score)
+
+    async def _aretrieve_from_object(
+        self,
+        obj: Any,
+        query_bundle: QueryBundle,
+        score: float,
+        *,
+        query_type: KQ,
+    ) -> List[NodeWithScore]:
+        """Retrieve nodes from object."""
+        if isinstance(obj, type(self)):
+            if self._verbose:
+                print_text(
+                    f"Retrieving from object {obj.__class__.__name__} with query {query_bundle.query_str}\n",
+                    color="llama_pink",
+                )
+
+                return await obj.aretrieve_multi_modal(
+                    query_bundle, query_type=query_type
+                )
+
+        return await super()._aretrieve_from_object(obj, query_bundle, score)
+
+    def _handle_recursive_retrieval(
```

Comment:
Unlike in `MultiModalVectorIndexRetriever`, composite nodes are nominally supported for non-text modalities, but this feature has yet to be tested.